Sources/Controllers/TargetMenu/GPX/TrackMenu/OATrackMenuHudViewController.mm

```diff
@@ -2153,14 +2727,123 @@ - (void)onSwitchPressed:(id)sender
+    values[@"selected_index_int_value"] = @(segment.selectedSegmentIndex);
+    [cellData setData:@{ kTableValues: values }];
+
+    if (cellData.updateData)
+        cellData.updateData();
+
+    [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:[cellData.values[@"statistics_row_int_value"] intValue]
+                                                                inSection:indexPath.section]]
+                          withRowAnimation:UITableViewRowAnimationNone];
+}
+
+- (void)onBarChartScrolled:(UIPanGestureRecognizer *)recognizer
+             lineChartView:(LineChartView *)lineChartView
+{
+    if (recognizer.state == UIGestureRecognizerStateChanged)
+    {
+        if (lineChartView.lowestVisibleX > 0.1
+                && [self getRoundedDouble:lineChartView.highestVisibleX] != [self getRoundedDouble:lineChartView.chartXMax])
+        {
+            _lastTranslation = [recognizer translationInView:lineChartView];
+            return;
+        }
+
+        ChartHighlight *lastHighlighted = lineChartView.lastHighlighted;
+        CGPoint touchPoint = [recognizer locationInView:lineChartView];
+        CGPoint translation = [recognizer translationInView:lineChartView];
+        ChartHighlight *h = [lineChartView getHighlightByTouchPoint:CGPointMake(lineChartView.isFullyZoomedOut
+                ? touchPoint.x : _highlightDrawX + (_lastTranslation.x - translation.x), 0.)];
+
+        if (h != lastHighlighted)
+        {
+            lineChartView.lastHighlighted = h;
+            [lineChartView highlightValue:h callDelegate:YES];
+        }
+    }
+    else if (recognizer.state == UIGestureRecognizerStateEnded)
+    {
+        _lastTranslation = CGPointZero;
+        if (lineChartView.highlighted.count > 0)
+            _highlightDrawX = lineChartView.highlighted.firstObject.drawX;
+    }
+}
+
+- (void)onChartGesture:(UIGestureRecognizer *)recognizer
+         lineChartView:(LineChartView *)lineChartView
+{
+    if (recognizer.state == UIGestureRecognizerStateBegan)
+    {
+        _hasTranslated = NO;
+        if (lineChartView.highlighted.count > 0)
+            _highlightDrawX = lineChartView.highlighted.firstObject.drawX;
+        else
+            _highlightDrawX = -1;
+    }
+    else if (([recognizer isKindOfClass:UIPinchGestureRecognizer.class] ||
+            ([recognizer isKindOfClass:UITapGestureRecognizer.class]
+                    && (((UITapGestureRecognizer *) recognizer).nsuiNumberOfTapsRequired == 2)))
+            && recognizer.state == UIGestureRecognizerStateEnded)
+    {
+        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:lineChartView.tag & 0x3FF inSection:lineChartView.tag >> 10];
+        OAGPXTableCellData *cellData = [self getCellData:indexPath];
+
+        [_routeLineChartHelper refreshHighlightOnMap:YES
+                                           lineChartView:lineChartView
+                                        trackChartPoints:cellData.values[@"points_value"]];
+    }
```

Comment:
I think this class should be refactored and split into several classes. You should have one class that handles the UI and other classes that have the data for the UI (the model) and perhaps classes that control custom behaviours for different screens (like chart scrolling etc.). 3000 for a UI class is way too much.